demos/entrypoints/main.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"context"
+	"path/filepath"
+
+	"dagger.io/dagger"
+)
+
+func main() {
+	dagger.Serve(
+		Build,
+		Test,
+	)
+}
+
+// Build the go binary from the given go repo, branch, and subpath. Silly right now due to lack of host export, but proves concept
+func Build(ctx context.Context, repo, branch, subpath string) (string, error) {
+	c, err := dagger.Connect(ctx)
+	if err != nil {
+		return "", err
+	}
+	defer c.Close()
+
+	binpath := filepath.Join("./bin", filepath.Base(subpath))
+	return baseContainer(c, repo, branch).
+		WithExec([]string{"go", "build", "-v", "-x", "-o", binpath, "./" + subpath}).
+		Stderr(ctx)
+}
+
+// Test the go binary from given go repo, branch, and subpath
+func Test(ctx context.Context, repo, branch, subpath string) (string, error) {
+	c, err := dagger.Connect(ctx)
+	if err != nil {
+		return "", err
+	}
+	defer c.Close()
```

Comment:
Have you thought about of having each of these magically take a `*dagger.Client` arg? Seems like it could make the DX more seamless, saving 4 lines per function, which in some cases is >50% of the function.

You wouldn't be able to pass different options to each client anymore, but I wonder if we could go down a path where you just don't need to set those options anymore, and it seems like at least some of them could be configured unilaterally across all entrypoints (e.g. log output).

Reply:
Oh another idea along this vein before I forget, we could also have those `dagger.Client`s be pre-populated with Pipeline settings that match the name of the entrypoint (including hierarchies of them when invoking subcommands)

Reply:
> Oh another idea along this vein before I forget, we could also have those `dagger.Client`s be pre-populated with Pipeline settings that match the name of the entrypoint (including hierarchies of them when invoking subcommands)

Yes. Please. I love this idea!

Reply:
Part of my latest updates today included `dagger.Context` with a `.Client()` method. I didn't get around to trying to add a pipeline labels automatically yet, but that should be pretty straightforward.

Also need to think about what the equivalent of this would be in Python. At least the magic pipeline labeling should also work there too, though the need to reduce client creation boilerplate probably doesn't apply.

Reply:
> Also need to think about what the equivalent of this would be in Python. 

May I suggest checking out how airbyte does it? They have a whole entrypoint system built on the Python SDK. Can’t hurt to look at how they pass the client around.

Reply:
> Also need to think about what the equivalent of this would be in Python.

I can help you there. When I think about doing a limited schema based on CLI, I imagine creating a high level interface that uses [typer](https://typer.tiangolo.com) (like Mage; we're using it for [running codegen](https://github.com/dagger/dagger/blob/51f6dd9c5c2e64e55ff497a48ce1886bbca77103/sdk/python/src/dagger/cli.py) and [extensions schema](https://github.com/dagger/dagger/blob/51f6dd9c5c2e64e55ff497a48ce1886bbca77103/sdk/python/src/dagger/server/cli.py)) and Strawberry underneath.

When you create a Typer app, you can create a context which is shared in subcommands. So a single dagger connection can be established in the beginning, including config, and shared.

Something like this:


```python
import dagger

app = dagger.Project()

@app.command()
async def build(ctx: dagger.Context, foo: str) -> str:
    base_image = ctx.client.container().from_("golang:1.20-alpine")
    ...

@app.command()
async def test(ctx: dagger.Context, bar: str) -> str:
    ...

if __name__ == "__main__"
    app()
```

Or using classes:

```python
import dagger

class Project(dagger.Project):
    @dagger.command()
    async def build(foo: str) -> str:
        base_image = self.client.container().from_("golang:1.20-alpine")
        ...

    @dagger.command()
    async def test(bar: str) -> str:
        ...

app = Project()

if __name__ == "__main__"
    app()
```

The last bit of boilerplate can be hidden away in the container runtime, but that implies having a fixed name for this variable:
```python
if __name__ == "__main__"
    app()
```

Reply:
@helderco I almost missed your comment here, that looks great! I was planning on asking your eventually about whether you'd think it'll make sense to create our own thin layer on top of strawberry to enable more dagger-specific niceties like this (while still being able to re-use strawberry underneath the hood as much as possible), sounds like the answer is yes :-)

Reply:
+1 to removing boilerplate if we can!